ipfs/provider.go

```diff
@@ -4,7 +4,44 @@ import (
 	"io"
 
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/options"
 )
 
 // APIProvider allows customizable IPFS core APIs.
-type APIProvider func() (coreiface.APIDagService, io.Closer, error)
+type APIProvider func(online bool) (coreiface.APIDagService, io.Closer, error)
+
+type onlineOfflineProvider struct {
+	online, offline coreiface.APIDagService
+	closer          io.Closer
+}
+
+func NewOnlineOfflineProvider(online coreiface.CoreAPI, closer io.Closer) (APIProvider, error) {
+
+	offline, err := online.WithOptions(OfflineOption)
+	if err != nil {
+		return nil, err
+	}
+
+	provider := onlineOfflineProvider{
+		online:  online.Dag(),
+		offline: offline.Dag(),
+	}
+
+	return provider.Provide, nil
+}
+
+func (o onlineOfflineProvider) Provide(online bool) (coreiface.APIDagService, io.Closer, error) {
+	switch online {
+	case true:
+		return o.online, o.closer, nil
+	case false:
+		return o.offline, o.closer, nil
+	}
+	// the compiler does't know this code is unreachable
+	return nil, nil, nil
+}
+
+func OfflineOption(settings *options.ApiSettings) error {
+	settings.Offline = true
+	return nil
+}
```

Comment:
this allows us to still be able to use a dag only mock, and have an offline APIDagService for the blockstore

Reply:
Sorry that I am responding to this quite late, but I am against this way of having offline DAG. Its current usage is confusing, it feels like we have two IPFSs offline and online instead of one and provide two of them independently.

Instead, I would suggest going this way:
```go
func OfflineDAG(bstore blockstore.Blockstore) format.DAGService {
	return dag.NewDAGService(bsrv.New(bstore, offline.Exchange(bstore)))
}
```

or straightly the constructor
```go
func NewBlockStore(db dbm.DB, bstore blockstore.Blockstore) *BlockStore {
	bs := LoadBlockStoreState(db)
	return &BlockStore{
		base:   bs.Base,
		height: bs.Height,
		db:     db,
		dag:    dag.NewDAGService(bsrv.New(bstore, offline.Exchange(bstore))),
	}
}
```

And, importantly continue our road of decomposing IPFS into pieces, specifically, here we would already rely on IPFS's blockstore instead of very high-level API. Besides this, we would also simply some testability with easily mockable Blocksktore

Reply:
all bueno, it's better that we get it right the first time! :slightly_smiling_face:  I like this better too, it's less hacky and ensures that all `store.Blockstore`s are offline. The thing that I think can be improved is readability. I can see this being incredibly confusing to someone unfamiliar with what we're doing. The word blockstore is used way too much, and the purpose isn't directly obvious. It will get even more confusing when we have to modify the `APIProvider` to also return a `blockstore.Blockstore`, which we pass into the `store.Blockstore` to make a dag, which is also returned by the `APIProvider`, but we only use that dag api for consensus... We know it's because we want to enforce the online and offline usage of the DAG API, but I think we could make that more readable by doing something like this

```go
// Maybe name this something else, like Gateway or Connector
// APIProvider allows customizable IPFS core APIs.
type APIProvider func() (DAGProvider, io.Closer, error)

// DAGProvider returns an online or offline DAG API
type DAGProvider func(online bool) coreiface.APIDagService
```
and
```go
func NewBlockStore(db dbm.DB, dagProvider ipfs.DAGProvider) *BlockStore {
	bs := LoadBlockStoreState(db)
	return &BlockStore{
		base:   bs.Base,
		height: bs.Height,
		db:     db,
		dag:    dagProvider(false),
	}
}
```

Reply:
we should for sure use the mechanism you mentioned 
```go
func OfflineDAG(bstore blockstore.Blockstore) format.DAGService {
	return dag.NewDAGService(bsrv.New(bstore, offline.Exchange(bstore)))
}
```
in the `DAGProvider` instead of the one used in `onlineOfflineProvider`

Reply:
I guess we could also do something like this too
```go
// maybe name this something different
// APIProvider allows customizable IPFS core APIs.
type APIProvider func() (exchange.Interface, blockstore.Blockstore, io.Closer, error)
```
and just construct the `DAGServices` wherever needed

Reply:
or
```go
func GetAPI(node *core.IpfsNode) (exchange.Interface, blockstore.Blockstore, io.Closer) {
	return node.Exchange, node.Blockstore, node
}
```

Reply:
Take a look at the new PuBlock semantics [here](https://github.com/lazyledger/lazyledger-core/blob/hlib/zero-bootstrapping/p2p/ipld/write.go#L27). It now accepts ContentRouting from IPFS/libp2p.

Reply:
So overall, to keep readibiltiy, we can rename APIProvider to NodeProvider

Reply:
I'll give that PR a more thorough look over, and try and come up with a suitable solution over the weekend.

Reply:
what do you think about #400?